### PR TITLE
Add Eventing auth manager feature flag

### PR DESCRIPTION
**Pull request by marcobebway**
**Description**

Add Eventing auth manager feature flag.

**Related issue(s)**

- https://github.com/kyma-project/kyma/issues/17490

### CODE REVIEW

**Comment by friedrichwilken** [comment]
please keep this on hold until the rel cutoff for `2.15` is done.

File: components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go [hunk 1]
```diff
@@ -337,6 +336,209 @@ func Test_updateMutatingValidatingWebhookWithCABundle(t *testing.T) {
...
+				// not required secret
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      getOAuth2ClientSecretName(),
+						Namespace: deployment.ControllerNamespace,
+					},
+					Data: map[string][]byte{
+						secretKeyClientID:     []byte("test-client-id-1"),
+						secretKeyClientSecret: []byte("test-client-secret-1"),
+						secretKeyTokenURL:     []byte("test-token-url-1"),
```
Comment (muralov) [inline components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go:455]: nitpick: maybe delete token url as oauth2 secret doesn't have it normally

File: components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go [hunk 2]
```diff
@@ -337,6 +336,209 @@ func Test_updateMutatingValidatingWebhookWithCABundle(t *testing.T) {
...
+				// required secret
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      getOAuth2ClientSecretName(),
+						Namespace: deployment.ControllerNamespace,
+					},
+					Data: map[string][]byte{
+						secretKeyClientID:     []byte("test-client-id-1"),
+						secretKeyClientSecret: []byte("test-client-secret-1"),
+						secretKeyTokenURL:     []byte("test-token-url-1"),
```
Comment (muralov) [inline components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go:489]: nitpick: maybe delete token url as oauth2 secret doesn't have it normally

File: components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go [hunk 3]
```diff
@@ -337,6 +336,209 @@ func Test_updateMutatingValidatingWebhookWithCABundle(t *testing.T) {
...
+				// not required secret
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      getOAuth2ClientSecretName(),
+						Namespace: deployment.ControllerNamespace,
+					},
+					Data: map[string][]byte{
+						secretKeyClientID:     []byte("test-client-id-1"),
+						secretKeyClientSecret: []byte("test-client-secret-1"),
+						secretKeyTokenURL:     []byte("test-token-url-1"),
```
Comment (muralov) [inline components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go:393]: nitpick: maybe delete token url as oauth2 secret doesn't have it normally

File: components/eventing-controller/pkg/backend/eventmesh/eventmesh.go [hunk 4]
```diff
@@ -85,11 +85,11 @@ func (em *EventMesh) Initialize(cfg env.Config) error {
 
 // getWebHookAuth returns the webhook auth config from the given env config
 // or returns an error if the env config contains invalid grant type or auth type.
-func getWebHookAuth(cfg env.Config, credentials *OAuth2ClientCredentials) *types.WebhookAuth {
+func getWebHookAuth(credentials *OAuth2ClientCredentials) *types.WebhookAuth {
 	return &types.WebhookAuth{
 		ClientID:     credentials.ClientID,
 		ClientSecret: credentials.ClientSecret,
-		TokenURL:     cfg.WebhookTokenEndpoint,
+		TokenURL:     credentials.TokenURL,
```
Comment (muralov) [inline components/eventing-controller/pkg/backend/eventmesh/eventmesh.go:92]: hope `credentials.TokenURL` is set for both cases (feature flag enabled & disabled)

File: components/eventing-controller/controllers/backend/reconciler.go [hunk 5]
```diff
@@ -703,23 +713,54 @@ func (r *Reconciler) getCurrentBackendCR(ctx context.Context) (*eventingv1alpha1
...
-		err = errors.New("key 'client_id' not found in secret " + oauth2SecretNamespacedName.String())
-		return
+	if clientID, exists = oauth2Secret.Data[secretKeyClientID]; !exists {
+		err = errors.Errorf("key '%s' not found in secret %s",
+			secretKeyClientID, oauth2SecretNamespacedName.String())
+		return clientID, clientSecret, tokenURL, err
+	}
+	if clientSecret, exists = oauth2Secret.Data[secretKeyClientSecret]; !exists {
+		err = errors.Errorf("key '%s' not found in secret %s",
+			secretKeyClientSecret, oauth2SecretNamespacedName.String())
```
Comment (muralov) [inline components/eventing-controller/controllers/backend/reconciler.go:750]: we shouldn't log secret data

File: components/eventing-controller/controllers/backend/reconciler.go [hunk 6]
```diff
@@ -793,9 +834,13 @@ func (r *Reconciler) stopNATSController() error {
...
-func (r *Reconciler) startBEBController(clientID, clientSecret []byte) error {
+func (r *Reconciler) startBEBController(clientID, clientSecret, tokenURL []byte) error {
 	if !r.bebSubMgrStarted {
-		bebSubMgrParams := subscriptionmanager.Params{"client_id": clientID, "client_secret": clientSecret}
+		bebSubMgrParams := subscriptionmanager.Params{
+			subscriptionmanager.ParmaNameClientID:     clientID,
+			subscriptionmanager.ParmaNameClientSecret: clientSecret,
+			subscriptionmanager.ParmaNameTokenURL:     tokenURL,
+		}
 		if err := r.bebSubMgr.Start(r.cfg.DefaultSubscriptionConfig, bebSubMgrParams); err != nil {
```
Comment (muralov) [inline components/eventing-controller/controllers/backend/reconciler.go:844]: hope `bebSubMgrParams` are not printed anywhere

**Review by muralov — COMMENTED** [review]
We should not print some confidential data.

File: components/eventing-controller/controllers/backend/reconciler.go [hunk 7]
```diff
@@ -703,23 +713,54 @@ func (r *Reconciler) getCurrentBackendCR(ctx context.Context) (*eventingv1alpha1
...
+	}
+	if !featureflags.IsEventingWebhookAuthEnabled() {
+		tokenURL = []byte(r.envCfg.WebhookTokenEndpoint)
+		return clientID, clientSecret, tokenURL, err
 	}
-	if clientSecret, exists = oauth2Secret.Data["client_secret"]; !exists {
-		err = errors.New("key 'client_secret' not found in secret " + oauth2SecretNamespacedName.String())
-		return
+	if tokenURL, exists = oauth2Secret.Data[secretKeyTokenURL]; !exists {
+		err = errors.Errorf("key '%s' not found in secret %s",
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler.go:758]: We decided to keep it like this to be consistent with the rest of the code.

File: components/eventing-controller/controllers/backend/reconciler.go [hunk 8]
```diff
@@ -7,12 +7,9 @@ import (
 	"fmt"
 	"os"
 
-	pkgerrors "github.com/kyma-project/kyma/components/eventing-controller/pkg/errors"
-	"golang.org/x/xerrors"
-
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
-
+	"golang.org/x/xerrors"
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler.go:12]: I just re-organized the imports as required by the linter.

File: components/eventing-controller/controllers/backend/reconciler.go (hunk 5 above)
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler.go:750]: We don't, it is only the secret name and namespace.

File: components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go (hunk 1 above)
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go:455]: As agreed, we will leave it like that.

File: components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go (hunk 2 above)
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler_internal_unit_test.go:489]: As agreed, we will leave it like that.

File: components/eventing-controller/internal/featureflags/featureflags.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+// SetEventingWebhookAuthEnabled enable/disable the Eventing webhook auth feature flag.
+func SetEventingWebhookAuthEnabled(enabled bool) {
+	f.eventingWebhookAuthEnabled = enabled
+}
+
+// IsEventingWebhookAuthEnabled returns true if the Eventing webhook auth feature flag is enabled,
+// otherwise returns false.
+func IsEventingWebhookAuthEnabled() bool {
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/internal/featureflags/featureflags.go:19]: As agreed, we will leave it like that.

File: components/eventing-controller/pkg/subscriptionmanager/beb/beb.go [hunk 10]
```diff
@@ -244,19 +244,26 @@ func cleanupEventMesh(backend backendeventmesh.Backend, dynamicClient dynamic.In
 }
 
 func getOAuth2ClientCredentials(params subscriptionmanager.Params) (*backendeventmesh.OAuth2ClientCredentials, error) {
-	val := params["client_id"]
+	val := params[subscriptionmanager.ParmaNameClientID]
 	id, ok := val.([]byte)
 	if !ok {
-		return nil, fmt.Errorf("expected []byte value for client_id, but received %T", val)
+		return nil, fmt.Errorf("expected []byte value for %s, but received %T", subscriptionmanager.ParmaNameClientID, val)
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/pkg/subscriptionmanager/beb/beb.go:250]: Where are printing only the type not the value.

File: components/eventing-controller/pkg/subscriptionmanager/beb/beb.go [hunk 11]
```diff
@@ -244,19 +244,26 @@ func cleanupEventMesh(backend backendeventmesh.Backend, dynamicClient dynamic.In
...
 	}
-	val = params["client_secret"]
+	val = params[subscriptionmanager.ParmaNameClientSecret]
 	secret, ok := val.([]byte)
 	if !ok {
-		return nil, fmt.Errorf("expected []byte value for client_secret, but received %T", val)
+		return nil, fmt.Errorf("expected []byte value for %s, but received %T",
+			subscriptionmanager.ParmaNameClientSecret, val)
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/pkg/subscriptionmanager/beb/beb.go:256]: Where are printing only the type not the value.

File: components/eventing-controller/pkg/subscriptionmanager/beb/beb.go [hunk 12]
```diff
@@ -244,19 +244,26 @@ func cleanupEventMesh(backend backendeventmesh.Backend, dynamicClient dynamic.In
...
+	}
+	val = params[subscriptionmanager.ParmaNameTokenURL]
+	tokenURL, ok := val.([]byte)
+	if !ok {
+		return nil, fmt.Errorf("expected []byte value for %s, but received %T", subscriptionmanager.ParmaNameTokenURL, val)
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/pkg/subscriptionmanager/beb/beb.go:261]: Where are printing only the type not the value.

File: components/eventing-controller/controllers/backend/reconciler.go (hunk 6 above)
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler.go:848]: I did not introduce any logs that print it.

File: components/eventing-controller/pkg/backend/eventmesh/eventmesh.go (hunk 4 above)
Comment (marcobebway) [reply components/eventing-controller/pkg/backend/eventmesh/eventmesh.go:92]: It is set here https://github.com/kyma-project/kyma/blob/88da789a122950f526ea5243e813bcfb5320f266/components/eventing-controller/controllers/backend/reconciler.go#L732.

**Comment by marcobebway** [comment]
> We should not print some confidential data.

We don't print any confidential data. We only print the secret name/namespace.

File: components/eventing-controller/controllers/backend/reconciler.go [hunk 13]
```diff
@@ -703,23 +713,54 @@ func (r *Reconciler) getCurrentBackendCR(ctx context.Context) (*eventingv1alpha1
...
+		name = r.cfg.EventingWebhookAuthSecretName
+		namespace = r.cfg.EventingWebhookAuthSecretNamespace
+	} else {
+		name = getOAuth2ClientSecretName()
+		namespace = deployment.ControllerNamespace
+	}
+	return types.NamespacedName{Name: name, Namespace: namespace}
+}
+
+func (r *Reconciler) getOAuth2ClientCredentials(ctx context.Context) ([]byte, []byte, []byte, error) {
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler.go]: Created `oauth2Credentials` struct.

File: components/eventing-controller/controllers/backend/reconciler.go [hunk 14]
```diff
@@ -793,9 +834,13 @@ func (r *Reconciler) stopNATSController() error {
 	return nil
 }
 
-func (r *Reconciler) startBEBController(clientID, clientSecret []byte) error {
+func (r *Reconciler) startBEBController(clientID, clientSecret, tokenURL []byte) error {
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler.go]: Done

**Review by muralov — DISMISSED** [review]

File: components/eventing-controller/controllers/backend/reconciler.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/eventing-controller/controllers/backend/reconciler.go]: Done

**Review by muralov — APPROVED** [review]

**Pull request merged**